Source/Core/Common/SPSCQueue.h

```diff
@@ -6,104 +6,181 @@
 {
 public:
-  SPSCQueue() : m_size(0) { m_write_ptr = m_read_ptr = new ElementPtr(); }
-  ~SPSCQueue()
+  std::size_t Size() const { return m_size.load(std::memory_order_acquire); }
+
+protected:
+  void IncSize() { m_size.fetch_add(1, std::memory_order_release); }
+  void DecSize() { m_size.fetch_sub(1, std::memory_order_release); }
+
+  std::atomic<std::size_t> m_size = 0;
+};
+
+#if defined(ATOMIC_NOTIFY_ONE_IS_UNAVAILABLE)
+class WaitableAtomicSize : public AtomicSize
+{
+public:
+  void WaitForEmpty()
   {
-    // this will empty out the whole queue
-    delete m_read_ptr;
+    std::unique_lock lg{m_lock};
+    m_cond.wait(lg, [this] { return Size() == 0; });
+  }
+
+  void WaitForData()
+  {
+    std::unique_lock lg{m_lock};
+    m_cond.wait(lg, [this] { return Size() != 0; });
+  }
+
+protected:
+  void IncSize()
+  {
+    std::unique_lock lg{m_lock};
+    AtomicSize::IncSize();
+    m_cond.notify_one();
   }
 
-  u32 Size() const
+  void DecSize()
+  {
+    std::unique_lock lg{m_lock};
+    AtomicSize::DecSize();
+    m_cond.notify_one();
+  }
+
+private:
+  std::condition_variable m_cond;
+  std::mutex m_lock;
+};
+#else
+class WaitableAtomicSize : public AtomicSize
+{
+public:
+  void WaitForEmpty()
+  {
+    while (auto const old_size = Size())
+      m_size.wait(old_size, std::memory_order_acquire);
+  }
+
+  void WaitForData() { m_size.wait(0, std::memory_order_acquire); }
+
+protected:
+  void IncSize()
   {
-    static_assert(NeedSize, "using Size() on SPSCQueue without NeedSize");
-    return m_size.load();
+    AtomicSize::IncSize();
+    m_size.notify_one();
   }
 
-  bool Empty() const { return !m_read_ptr->next.load(); }
-  T& Front() const { return m_read_ptr->current; }
-  template <typename Arg>
-  void Push(Arg&& t)
+  void DecSize()
   {
-    // create the element, add it to the queue
-    m_write_ptr->current = std::forward<Arg>(t);
-    // set the next pointer to a new element ptr
-    // then advance the write pointer
-    ElementPtr* new_ptr = new ElementPtr();
-    m_write_ptr->next.store(new_ptr, std::memory_order_release);
+    AtomicSize::DecSize();
+    m_size.notify_one();
+  }
+};
+#endif
+};  // namespace detail
+
+template <typename T, typename SizeBase = detail::AtomicSize>
+class SPSCQueue final : public SizeBase
+{
+public:
+  SPSCQueue() = default;
+  ~SPSCQueue()
+  {
+    Clear();
+    delete m_read_ptr;
+  }
+
+  SPSCQueue(const SPSCQueue&) = delete;
+  SPSCQueue& operator=(const SPSCQueue&) = delete;
+
+  bool Empty() const { return SizeBase::Size() == 0; }
+
+  // The following are only safe from the "producer thread":
+  void Push(const T& arg) { Emplace(arg); }
+  void Push(T&& arg) { Emplace(std::move(arg)); }
+  template <typename... Args>
+  void Emplace(Args&&... args)
+  {
+    std::construct_at(&m_write_ptr->value.data, std::forward<Args>(args)...);
+
+    Node* const new_ptr = new Node;
+    m_write_ptr->next = new_ptr;
     m_write_ptr = new_ptr;
-    if (NeedSize)
-      m_size++;
+
```

Comment:
Well, I fully admit that lock-free multithreading logic is not my forte, so explain me the details here. All the synchronization here goes through the `SizeBase::IncSize()`, right? That's because `m_size.fetch_add(1, std::memory_order_release);` guarantees that by the time the other thread sees the increased size the writes to `m_write_ptr->next` and `m_write_ptr` have completed, correct?